rfd/0096-securing-docker-images.md

```diff
@@ -0,0 +1,378 @@
+> In the real distribution there would be multiple packages required.
+
+### A digression: Alternative builders
+
+As part of researching this RFD, I examined a couple of alternative ways to construct 
+the Teleport image. 
+
+ * **bazel, [distroless](https://github.com/GoogleContainerTools/distroless) and [rules_docker](https://github.com/bazelbuild/rules_docker)**: 
+   Given that the underlying distroless images are built using `bazel`, it should be 
+   possible to construct a custom image for Teleport in the same way. After some 
+   experimentation, I found that
+
+   1. the Debian package installation technique used by `rules_docker` is essentially 
+      a tweaked version of the extract-and-copy approach used by the `Dockerfile` 
+      above, and
+
+   2. There is a major [chicken-or-egg problem](https://github.com/GoogleContainerTools/distroless/issues/542)
+      in the build process when `distroless` is used as an external dependency in 
+      an enclosing `bazel` workspace, requiring manual intervention in the build to
+      solve.
+
+   Using `bazel` does not resolve a major limitation of using a basic 
+   `Dockerfile` (i.e. the `xcopy` style install) and introduces more complexity, in 
+   terms of both build process and tooling, so was rejected in favour of the 
+   `Dockerfile` approach.
+
+ * **[apko](https://github.com/chainguard-dev/apko)**: Apko is a tool for quickly 
+   building minimalist, reproducible Alpine linux images, using a declarative 
+   format. While I found the tool to be very neat, the images it generates are 
+   still closer to a "debug" Distroless image. 
+
+   Using `apko` would also require us to build an Alpine linux package for Teleport 
+   to integrate it nto the build.
+
+   I seriously considered recommending `apko`, as it has some neat features (e.g.
+   automatically producting a SBOM as part of the construction process), but in the 
+   end I rejected it because of the extra software included in the resulting images.
+
+### Smoke testing
+
+A smoke test is a simple, quick test to assert basic functionality. We simply want to 
+find out if Teleport will even start in the environment contained by our container 
+image. While this is not _per se_ a security matter, using a distroless base image 
+means we have very little padding if an unexpected dependency finds its way into
+Teleport. 
+
+We want to stop ourselves shipping garbage to our customers if at all possible.
+
+### Image signing
+
+In order to allow our customers to validate our published Teleport images, 
+our images will be signed using the `cosign` tool, similarly to the Distroless 
+base images.
+
+The `cosign` tool [integrates well with GHA](https://github.blog/2021-12-06-safeguard-container-signing-capability-actions/), 
+and is even included in the template "how to publish a docker image" example 
+workflow. It also provides "keyless" signing via OIDC, meaning that our build
```

Comment:
@tcsc As part of our automatic upgrades work we want our Kube Operator to be able to pull down an image and verify its signature. According to the documentation, it seems like keyless signing requires the use of OIDC?

https://docs.sigstore.dev/cosign/sign/#keyless-signing

> You can use Cosign to sign without keys by authenticating with an OIDC (OpenID Connect) protocol supported by Sigstore. Currently, you can authenticate with Google, GitHub, or Microsoft.
> This will open a browser window to authenticate your credentials for the signature.

In which case the Operator wouldn't be able to do that? Unless I'm misunderstanding how it works.

Should we use signing with regular keys instead and host the public key somewhere Operator can fetch it and verify the signature?

cc @fspmarshall @hugoShaka @reedloden

Reply:
Looks like keyless signing verification doesn't require OIDC, just access to the public transparency log so operator should still be able to verify it.

Would be nice to try it out though and confirm - @tcsc do you have any signed images published anywhere so @hugoShaka can use them as examples as he's working on building the upgrade operator?

Reply:
The Kubernetes updater will be deployed once and should be able to operate for 1+ year without being updated. As Keyless signing is experimental and can know breaking changes in the near future, I think it is a bit risky to rely only on Keyless signing for such a long-lived piece of software.

Keyless is way less costly than setting up a secure signing system, I suggest we use it by default but still generate a failover signing key pair, store it somewhere where no one has access and bake it in the validation policies/binaries we're going to ship.

This way, if we have an issue with keyless, have to move out of GitHub actions, Fulcio is down, or face a major CI incident, we would be able to break the glass and start using the keypair instead of asking everyone to update their policy/validation mechanism.

This increases the attack surface as we would have an additional secret to secure, but I think the tradeoff could be positive if we consider the availability and resiliency guarantees we gain.

This might also address some of @wadells [concerns](https://github.com/gravitational/teleport/pull/19045/files#r1072712358).

Reply:
> This might also address some of @wadells [concerns](https://github.com/gravitational/teleport/pull/19045/files#r1072712358).

Yeah -- a key we control would certainly alleviate my future-proofing concern.  My experience is that Teleport CI/Release systems have a 3-4 year lifespan, and thus I'm hesitant to have a tight coupling between the build system and the customer validation UX.  I'd much prefer a design where we can plug the build infra without any change needed on the consumer side.